libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -334,46 +342,81 @@ export class WorkspaceInstallManager extends BaseInstallManager {
         );
       }
 
-      Utilities.executeCommandWithRetry(
-        {
-          command: packageManagerFilename,
-          args: installArgs,
-          workingDirectory: this.rushConfiguration.commonTempFolder,
-          environment: packageManagerEnv,
-          suppressOutput: false
-        },
-        this.options.maxInstallAttempts,
-        () => {
-          if (this.rushConfiguration.packageManager === 'pnpm') {
-            console.log(colors.yellow(`Deleting the "node_modules" folder`));
-            this.installRecycler.moveFolder(commonNodeModulesFolder);
-
-            // Leave the pnpm-store as is for the retry. This ensures that packages that have already
-            // been downloaded need not be downloaded again, thereby potentially increasing the chances
-            // of a subsequent successful install.
-
-            Utilities.createFolderWithRetry(commonNodeModulesFolder);
+      // Store the tip IDs that should be printed.
+      // They will be printed all at once *after* the install
+      const tipIDsShouldBePrinted: Set<CustomTipIdEnum> = new Set();
+      const supportedPnpmTips: ICustomTipInfo[] = Object.keys(CustomTipsConfiguration.CustomTipRegistry)
+        .map((key: string) => {
+          return CustomTipsConfiguration.CustomTipRegistry[key as CustomTipIdEnum];
+        })
+        .filter((tipInfo) => tipInfo.type === CustomTipType.pnpm);
+
+      const onPnpmStdoutChunk = (chunk: string): void => {
+        // Iterate over the supported custom tip metadata and try to match the chunk.
+        // TODO: this might be slow. Optimize it.
+        supportedPnpmTips.forEach((tipInfo: ICustomTipInfo) => {
+          if (tipInfo.isMatch && tipInfo.isMatch(chunk)) {
+            tipIDsShouldBePrinted.add(tipInfo.tipId);
           }
-        }
-      );
+        });
+      };
+
+      try {
+        await Utilities.executeCommandAndProcessOutputWithRetryAsync(
+          {
+            command: packageManagerFilename,
+            args: installArgs,
+            workingDirectory: this.rushConfiguration.commonTempFolder,
+            environment: packageManagerEnv,
+            suppressOutput: false
+          },
+          this.options.maxInstallAttempts,
+          onPnpmStdoutChunk,
+          () => {
+            if (this.rushConfiguration.packageManager === 'pnpm') {
+              console.log(colors.yellow(`Deleting the "node_modules" folder`));
```

Comment:
Is there a `Terminal` instance you can use here instead?

Reply:
Yes, I changed it to `this._terminal.writeWarningLine`. However, please note that this is beyond the scope of this PR. The callback logic was already present before. (But I guess there is no harm in using `terminal` to print)